File: examples/python/read.py
# Copyright (c) 2026 dingodb.com, Inc. All Rights Reserved
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""read.py — write a file then read it back, demonstrating seek/tell."""

import sys
from dingofs import DingofsError
from common import make_client

DATA = b"Hello DingoFS SDK!"

client = make_client()
try:
    client.mkdir("/demo_dir", 0o755)
    print("[ OK ] MkDir")

    # Write
    with client.open("/demo_dir/hello.txt", "wb") as f:
        f.write(DATA)
    print(f"[ OK ] Write  {len(DATA)} bytes")

    # Read back with seek/tell demo
    with client.open("/demo_dir/hello.txt", "rb") as f:
        # read all at once (size=-1 reads to EOF)
        content = f.read()
        print(f"[ OK ] Read   {len(content)} bytes")
        print(f"         content : [{content.decode()}]")
        print(f"         tell()  : {f.tell()}")

        # seek back to the beginning and read a partial chunk
        f.seek(0)
        print(f"[ OK ] Seek(0)   tell={f.tell()}")
        chunk = f.read(5)
        print(f"[ OK ] Read(5)   [{chunk.decode()}]  tell={f.tell()}")

        # seek from current position
        f.seek(1, 1)   # SEEK_CUR: skip 1 byte
        rest = f.read()
        print(f"[ OK ] Read rest [{rest.decode()}]")

    # Verify
    if content != DATA:
        print("[FAIL] content mismatch", file=sys.stderr)
        sys.exit(1)
    print("[ OK ] Content verified")

    client.unlink("/demo_dir/hello.txt")
    client.rmdir("/demo_dir")
    print("[ OK ] Cleanup")

except DingofsError as e:
    print(f"[FAIL] {e}", file=sys.stderr)
    sys.exit(1)
finally:
    client.stop()
    print("[ OK ] Stop")

File: examples/cpp/read.cc
/*
 * Copyright (c) 2026 dingodb.com, Inc. All Rights Reserved
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// read.cc — write a file then read it back via the C API.

#include <fcntl.h>

#include <cstring>
#include <iostream>
#include <string>

#include "common.h"

int main() {
  uintptr_t h = make_client();

  DINGOFS_CHECK(dingofs_mount(h, MDS_ADDRS, FS_NAME, MOUNT_POINT), "dingofs_mount");

  // Setup: create directory
  DINGOFS_CHECK(dingofs_mkdir(h, "/demo_dir", 0755), "dingofs_mkdir");

  // Write file
  int wfd = dingofs_open(h, "/demo_dir/hello.txt", O_WRONLY | O_CREAT | O_TRUNC,
                         0644);
  if (wfd < 0) {
    std::cerr << "[FAIL] dingofs_open(write): " << strerror(-wfd) << "\n";
    dingofs_umount(h);
    dingofs_delete(h);
    return 1;
  }

  const char* data = "Hello DingoFS SDK!";
  const size_t data_len = strlen(data);

  ssize_t wn = dingofs_write(h, wfd, data, data_len);
  if (wn < 0) {
    std::cerr << "[FAIL] dingofs_write: " << strerror(static_cast<int>(-wn))
              << "\n";
    dingofs_close(h, wfd);
    dingofs_umount(h);
    dingofs_delete(h);
    return 1;
  }
  std::cout << "[ OK ] dingofs_write  bytes=" << wn << "\n";

  DINGOFS_CHECK(dingofs_flush(h, wfd), "dingofs_flush");
  DINGOFS_CHECK(dingofs_close(h, wfd), "dingofs_close(write)");

  // Read back
  int rfd = dingofs_open(h, "/demo_dir/hello.txt", O_RDONLY, 0);
  if (rfd < 0) {
    std::cerr << "[FAIL] dingofs_open(read): " << strerror(-rfd) << "\n";
    dingofs_umount(h);
    dingofs_delete(h);
    return 1;
  }

  char buf[256] = {};
  ssize_t rn = dingofs_read(h, rfd, buf, sizeof(buf) - 1);
  if (rn < 0) {
    std::cerr << "[FAIL] dingofs_read: " << strerror(static_cast<int>(-rn))
              << "\n";
    dingofs_close(h, rfd);
    dingofs_umount(h);
    dingofs_delete(h);
    return 1;
  }
  std::cout << "[ OK ] dingofs_read  bytes=" << rn << "\n";

  std::string content(buf, static_cast<size_t>(rn));
  std::cout << "content: [" << content << "]\n";

  if (content != std::string(data, data_len)) {
    std::cerr << "[FAIL] content mismatch\n";
    dingofs_close(h, rfd);
    dingofs_umount(h);
    dingofs_delete(h);
    return 1;
  }
  std::cout << "[ OK ] content verified\n";

  DINGOFS_CHECK(dingofs_close(h, rfd), "dingofs_close(read)");

  // Cleanup
  DINGOFS_CHECK(dingofs_unlink(h, "/demo_dir/hello.txt"), "dingofs_unlink");
  DINGOFS_CHECK(dingofs_rmdir(h, "/demo_dir"), "dingofs_rmdir");

  DINGOFS_CHECK(dingofs_umount(h), "dingofs_umount");
  dingofs_delete(h);
  return 0;
}
